x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwkValidateUtil.java

```diff
@@ -68,35 +73,43 @@ static JwkSetLoader.JwksAlgs filterJwksAndAlgorithms(final List<JWK> jwks, final
      * @return True if JWT type and strength match both requirements for the signature algorithm.
      */
-    static boolean isMatch(final JWK jwk, final String algorithm) {
+    static boolean isMatch(final JWK jwk, final String algorithm, JwtUtil.TraceBuffer tracer) {
         try {
             if ((JwtRealmSettings.SUPPORTED_SIGNATURE_ALGORITHMS_HMAC.contains(algorithm)) && (jwk instanceof OctetSequenceKey jwkHmac)) {
                 final int bits = jwkHmac.size();
                 final int min = MACSigner.getMinRequiredSecretLength(JWSAlgorithm.parse(algorithm));
                 final boolean isMatch = bits >= min;
                 if (isMatch == false) {
-                    LOGGER.trace("HMAC JWK [" + bits + "] bits too small for algorithm [" + algorithm + "] minimum [" + min + "].");
+                    tracer.append("HMAC JWK [" + bits + "] bits too small for algorithm [" + algorithm + "] minimum [" + min + "].");
                 }
                 return isMatch;
             } else if ((JwtRealmSettings.SUPPORTED_SIGNATURE_ALGORITHMS_RSA.contains(algorithm)) && (jwk instanceof RSAKey jwkRsa)) {
                 final int bits = JwkValidateUtil.computeBitLengthRsa(jwkRsa.toPublicKey());
                 final int min = RSAKeyGenerator.MIN_KEY_SIZE_BITS;
                 final boolean isMatch = bits >= min;
                 if (isMatch == false) {
-                    LOGGER.trace("RSA JWK [" + bits + "] bits too small for algorithm [" + algorithm + "] minimum [" + min + "].");
+                    tracer.append("RSA JWK [" + bits + "] bits too small for algorithm [" + algorithm + "] minimum [" + min + "].");
                 }
                 return isMatch;
             } else if ((JwtRealmSettings.SUPPORTED_SIGNATURE_ALGORITHMS_EC.contains(algorithm)) && (jwk instanceof ECKey jwkEc)) {
                 final Curve curve = jwkEc.getCurve();
                 final Set<Curve> allowed = Curve.forJWSAlgorithm(JWSAlgorithm.parse(algorithm));
                 final boolean isMatch = allowed.contains(curve);
                 if (isMatch == false) {
-                    LOGGER.trace("EC JWK [" + curve + "] curve not allowed for algorithm [" + algorithm + "] allowed " + allowed + ".");
+                    tracer.append("EC JWK [" + curve + "] curve not allowed for algorithm [" + algorithm + "] allowed " + allowed + ".");
                 }
                 return isMatch;
             }
         } catch (Exception e) {
-            LOGGER.trace("Unexpected exception", e);
+            Supplier<String> message = () -> format(
+                "Unexpected exception while matching JWK with kid [%s] to it's algorithm requirement.",
+                jwk.getKeyID()
+            );
+            if (logger.isTraceEnabled()) {
+                logger.trace(message.get(), e);
+            } else {
+                logger.debug(message.get());
+            }
         }
         return false;
```

Comment:
Since we trace log for each case when `isMatch == false`, should we also trace log here?

Reply:
I wanted debug to ensure something is logged at debug scope but without the stacktrace. It is possible that a malformed JWK could emit a stack trace for every authenticate request.